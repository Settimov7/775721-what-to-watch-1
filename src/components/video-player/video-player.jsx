import * as React from 'react';
import * as PropTypes from 'prop-types';

export class VideoPlayer extends React.PureComponent {
  constructor(props) {
    super(props);

    this._videoRef = React.createRef();

    this.state = {
      isPlaying: false,
    };
  }

  componentDidUpdate() {
    const video = this._videoRef.current;

    this._updateIsPlayingFromProps();

    if (this.state.isPlaying) {
      video.play();
    } else {
      video.load();
    }
  }

  render() {
    const {videoSrc, posterSrc, options} = this.props;
    const {width, height, isMuted = true, isLoop = true} = options;

    return (
      <video
        ref={this._videoRef}
        src={videoSrc}
        poster={posterSrc}
        width={width}
        height={height}
        loop={isLoop}
        muted={isMuted}
      />
    );
  }

  _updateIsPlayingFromProps() {
    const {isPlaying} = this.props;

    this.setState({
      isPlaying,
    });
  }
}

VideoPlayer.propTypes = {
  videoSrc: PropTypes.string.isRequired,
  posterSrc: PropTypes.string,
  options: PropTypes.shape({
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    isMuted: PropTypes.bool,
    isLoop: PropTypes.bool,
  }).isRequired,
  isPlaying: PropTypes.bool,
};
