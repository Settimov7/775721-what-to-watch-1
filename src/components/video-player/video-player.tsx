import * as React from 'react';

import {VideoOptions} from "../../types";

interface Props {
  isPlaying?: boolean,
  videoSrc: string,
  posterSrc: string
  options: VideoOptions,
}

interface State {
  isPlaying: boolean,
}

export class VideoPlayer extends React.PureComponent<Props, State> {
  private _videoRef: React.RefObject<HTMLVideoElement>;

  constructor(props) {
    super(props);

    this._videoRef = React.createRef();

    this.state = {
      isPlaying: false,
    };
  }

  componentDidUpdate() {
    const video = this._videoRef.current;

    this._updateIsPlayingFromProps();

    if (this.state.isPlaying) {
      video.play();
    } else {
      video.load();
    }
  }

  render() {
    const {videoSrc, posterSrc, options} = this.props;
    const {width, height, isMuted = true, isLoop = true} = options;

    return (
      <video
        ref={this._videoRef}
        src={videoSrc}
        poster={posterSrc}
        width={width}
        height={height}
        loop={isLoop}
        muted={isMuted}
      />
    );
  }

  _updateIsPlayingFromProps() {
    const {isPlaying} = this.props;

    this.setState({
      isPlaying,
    });
  }
}
