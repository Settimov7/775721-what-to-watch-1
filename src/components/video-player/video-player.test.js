import renderer from "react-test-renderer";
import * as React from "react";

import {VideoPlayer} from "./video-player";

const mock = {
  film: {
    id: 2,
    title: `Bohemian Rhapsody`,
    posterSrc: `poster.jpg`,
    videoSrc: `video.mp4`,
  },
};

it(`Video player correctly renders`, () => {
  const props = {
    videoSrc: mock.film.videoSrc,
    posterSrc: mock.film.posterSrc,
    options: {
      width: 100,
      height: 100,
    }
  };

  const tree = renderer.create(<VideoPlayer {...props}/>).toJSON();

  expect(tree).toMatchSnapshot();
});
