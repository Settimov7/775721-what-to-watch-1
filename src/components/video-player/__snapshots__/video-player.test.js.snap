// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video player correctly renders 1`] = `
<video
  height={100}
  loop={true}
  muted={true}
  poster="poster.jpg"
  src="video.mp4"
  width={100}
/>
`;
